src/de.rs

```diff
@@ -195,6 +254,116 @@ impl<'de> Deserializer<'de> {
         }
         Ok(len)
     }
+}
+
+impl<'de, R: Read> Deserializer<TeeReader<'de, R>> {
+    fn parse_vec(&mut self) -> Result<Vec<u8>> {
+        let len = self.parse_length()?;
+        let mut output = vec![0; len];
+        self.fill_slice(&mut output)?;
+        Ok(output)
+    }
+
+    fn parse_string(&mut self) -> Result<String> {
+        let vec = self.parse_vec()?;
+        String::from_utf8(vec).map_err(|_| Error::Utf8)
+    }
+}
+
+impl<'de, R: Read> BcsDeserializer<'de> for Deserializer<TeeReader<'de, R>> {
+    type MaybeBorrowedBytes = Vec<u8>;
+
+    fn fill_slice(&mut self, slice: &mut [u8]) -> Result<()> {
+        Ok(self.input.read_exact(slice)?)
+    }
+
+    fn parse_and_visit_str<V>(&mut self, visitor: V) -> Result<V::Value>
+    where
+        V: Visitor<'de>,
+    {
+        visitor.visit_string(self.parse_string()?)
+    }
+
+    fn parse_and_visit_bytes<V>(&mut self, visitor: V) -> Result<V::Value>
+    where
+        V: Visitor<'de>,
+    {
+        visitor.visit_byte_buf(self.parse_vec()?)
+    }
+
+    fn next_key_seed<K: DeserializeSeed<'de>>(
+        &mut self,
+        seed: K,
+    ) -> Result<(K::Value, Self::MaybeBorrowedBytes), Error> {
+        self.input.capture_buffer = Some(Vec::new());
```

Comment:
Interestingly, this code won't work if a map is itself keyed by a map like `BTreeMap<BTreeMap<u8, u8>, u32>`
(which does compile, see https://gist.github.com/rust-play/4f4819775a567501e2dfd8e41699bbce )

We could easily return a error, say `UnsupportedMapInKeyposition` in the case where `self.input.capture_buffer.is_some()` before line 298, but I believe that there is a simple fix:
* Make `capture_buffer` a vector of bytes. Let's rename it in the process, say `captured_keys: Vec<Vec<u8>>`
* Add an empty vector at the end to start recording a new key.
* When the current key is finished, pop the vector to take the bytes, but do not forget to also append the bytes at the end of the previous (still ongoing) recorded key, if any.

I made a branch pursuing this idea here: https://github.com/zefchain/bcs/pull/7

@preston-evans98 I could be missing something, so you help to finalize this PR is still much appreciated: can you pick the suggested changes, add a unit test for `BTreeMap<BTreeMap<u8, u8>, u32>`, then update this PR?

Thank you!

Reply:
Very good catch! Here's a quick test case to detect the issue: 

```

#[test]
fn test_nested_map() {
    use std::collections::BTreeMap as Map;

    let mut m = Map::new();
    m.insert(Map::from_iter([(1u8, 0u8); 5]), 3);
    let bytes = to_bytes(&m).unwrap();

    assert_eq!(from_bytes(&bytes), Ok(m));
    assert_eq!(from_bytes_via_reader(&bytes), Ok(m));
}
```

Working on a fix now.

Reply:
Your changes look good to me! I also tested that they worked with an additional layer of nesting. Pushing the changes for review 👍

Reply:
great